### ISSUE TITLE
Network Security Configuration in Android is not honouring <base-config> and fatal crash when using IP addresses as <domain>

### ISSUE THREAD

**Issue by sarbagyastha** [flutter/flutter#65841]
### A) Not honouring `<base-config>`
1. Update `network_security_config.xml` as
```xml
<?xml version="1.0" encoding="utf-8"?>
<network-security-config>
    <base-config cleartextTrafficPermitted="true"/>
    <domain-config cleartextTrafficPermitted="false">
        <!-- List the domains for which you want ensure HTTPS traffic -->
        <domain>secure.example.com</domain>
        <domain>dev.example.com</domain>
    </domain-config>
</network-security-config>
```
2. Set base URL with `http` scheme. _For instance http://10.0.2.2_
3. Run the app.

**Expected results:**
All the domains & IPs expect `example.com` & `dev.example.com` should be allowed to have `http` scheme.

**Actual results:**
`Insecure socket connections are disallowed by platform:` for all domains & IPs

### B) Fatal Crash when setting IP addresses as `<domain>`
1. Update `network_security_config.xml` as
```xml
<?xml version="1.0" encoding="utf-8"?>
<network-security-config>
    <base-config cleartextTrafficPermitted="false"/>
    <domain-config cleartextTrafficPermitted="true">
        <!-- List the domains for which you want opt out ensuring HTTPS traffic -->
        <domain>10.0.2.2</domain>
        <domain>insecure.example.com</domain>
    </domain-config>
</network-security-config>
```
2. Set base URL with `http` scheme. _For instance http://10.0.2.2_
3. Run the app.

**Expected results:**
IPs and domains listed above should be allowed to have `http` scheme.

**Actual results:**
![image](https://user-images.githubusercontent.com/9782008/93199453-1fc9bf00-f76e-11ea-98a6-f72671cdbbe7.png)

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel dev, 1.22.0-12.0.pre, on Linux, locale en_US.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
[✓] Chrome - develop for the web
[✓] Linux toolchain - develop for Linux desktop
[✓] Android Studio (version 4.0)
[✓] Connected device (5 available)
```

**Comment by pedromassangocode** [comment]
Hi @sarbagyastha 
Can you please provide your `flutter doctor -v`, your `flutter run --verbose` and a minimal complete reproducible code sample and clear steps in how to reproduce this issue?
Thank you

**Comment by sarbagyastha** [comment]
@pedromassango I guess the information provided above should be enough as the issue not at my side, but at the engine itself.

**Comment by pedromassangocode** [comment]
I don't think so. I'm part of the triage team and without knowing much about the engine I should be able to reproduce the issue from the steps provided by the author.

For example I'm not sure where to add `network_security_config.xml` in a Flutter project an where to do this:

> 1. Set base URL with `http` scheme. _For instance http://10.0.2.2_

Please provide your `flutter run --verbose` and a [minimal complete reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example) and clear steps in how to reproduce this issue?
Thank you

**Comment by sarbagyastha** [comment]
@pedromassangocode [Here the link to the repo with reproducible example ](https://github.com/sarbagyastha/network_config_app)

The configuration xml can be [found here](https://github.com/sarbagyastha/network_config_app/tree/master/android/app/src/main/res/xml).

Url can be set from [here](https://github.com/sarbagyastha/network_config_app/blob/1c5ef045d462628c2301e617c328f7c82d1ca8dd/lib/main.dart#L56).

And here is the `flutter doctor -v` output:
```
[✓] Flutter (Channel master, 1.22.0-10.0.pre.178, on Linux, locale en_US.UTF-8)
    • Flutter version 1.22.0-10.0.pre.178 at
      /home/username/Development/flutter_master
    • Framework revision ca2ecf7536 (56 minutes ago), 2020-09-15 10:47:03 -0400
    • Engine revision c6121c5d51
    • Dart version 2.10.0 (build 2.10.0-124.0.dev)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /home/username/Android/Sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /home/username/Development/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[✓] Android Studio (version 4.0)
    • Android Studio at /home/username/Development/android-studio
    • Flutter plugin version 50.0.2-dev.1
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] Connected device (4 available)
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86    • Android 10 (API
      29) (emulator)
    • Linux (desktop)         • linux         • linux-x64      • Linux
    • Web Server (web)        • web-server    • web-javascript • Flutter Tools
    • Chrome (web)            • chrome        • web-javascript • Google Chrome
      84.0.4147.105

• No issues found!
```

**Comment by mehmetf** [comment]
1. `<base-config>` is not being handled today.  That is an issue and we will fix it. For now, you can use https://developer.android.com/guide/topics/manifest/application-element#usesCleartextTraffic to set global policy.

2. Rejecting IP addresses was a deliberate decision. Android documentation does not ban IP addresses but does not explicitly permit them either. Android also has odd behavior where it is possible to include sub-domains of an IP address which does not make sense. iOS documentation specifically mentions not to use IP addresses:

![image](https://user-images.githubusercontent.com/6687929/93234537-43badf80-f731-11ea-8d29-a50e8ea87005.png)

**Comment by sarbagyastha** [comment]
> Rejecting IP addresses was a deliberate decision. Android documentation does not ban IP addresses but does not explicitly permit them either. Android also has odd behavior where it is possible to include sub-domains of an IP address which does not make sense. iOS documentation specifically mentions not to use IP addresses

@mehmetf iOS has `NSAllowsLocalNetworking` to enable local IPs though. But how are we going to do the same in Android, if IP addresses are being rejected ?

Let's say I want to permit clear-text traffic only for selected local IPs but enforce https on all domain names. How can i achieve that ?

**Comment by mehmetf** [comment]
We don't support that right now. If you need to be able to access certain IPs via clear-text, you need to disable this feature via `NSAllowsArbitraryLoads` and `usesCleartextTraffic`.

I can take it as a feature request to support a set of IPs but there's really no way to do this in a cross-platform way since Android does not have the equivalent of `NSAllowsLocalNetworking` and iOS doesn't have the equivalent of "allowing select local IPs". When you set `NSAllowsLocalNetworking`, iOS allows all IP addresses (at least that's what the documentation implies).

**Comment by mehmetf** [comment]
The real solution to this is to swap the dart:io network stack with the platform equivalents (NSUrl and friends on iOS, OkHttp or Cronet on Android).

**Comment by pedromassangocode** [comment]
> I can take it as a feature request to support a set of IPs but there's really no way to do this in a cross-platform way since Android does not have the equivalent of `NSAllowsLocalNetworking` and iOS doesn't have the equivalent of "allowing select local IPs". When you set `NSAllowsLocalNetworking`, iOS allows all IP addresses (at least that's what the documentation implies).

Are we considering this as a feature request? If there is no way of doing this then I think we can close it then

**Comment by mehmetf** [comment]
I have been contemplating it @pedromassangocode . One way is to provide a Dart API to create a mutable policy that can be set. We discussed this at length with the Dart SDK team before and it was undesirable because mutable global state is a recipe for disaster and undermines security no matter how many locking mechanisms we provide.

I have two AIs for this issue:

1. Fix base-config handling. That's a bug.
2. Enhance the documentation I published on breaking changes that explains if users want to whitelist individual IP addresses, they should disable this feature entirely.

**Comment by MrCreeper1008** [comment]
For anyone bumping into this issue when trying to permit http connection to locally hosted servers, you can allow cleartext traffic only in debug mode on Android:
1. In `src/debug/AndroidManifest.xml`, add `xmlns:tools="http://schemas.android.com/tools"` to the root level `<manifest>` tag.
2. Then, add
```xml
<application
    android:usesCleartextTraffic="true"
    tools:node="merge" />
```  
inside `<manifest>`. The `tools:node="merge"` merges the `<application>` tag in the main AndroidManifest to this one.     Since debug AndroidManifest has a higher priority, properties specified here will override the ones specified in the main   manifest. More info [here (android developer website)](https://developer.android.com/studio/build/manifest-merge).

3. Run the application like normal

The app should now permit http connections but only if it is compiled in debug mode.

**Comment by daadu** [comment]
I have filed #77702, which I think is related.

Is it known that flutter does not respect `NSAllowsLocalNetworking` in `Info.Plist`

**Comment by abhi-kyros** [comment]
Have you found some solution in Android? As I am facing same.
